### PR TITLE
NIAD-1469: diagnostic report with no results

### PR DESCRIPTION
**Pull request by dawidskainos**

### CODE REVIEW

**Review by Broom94 — APPROVED** [review]

File: service/src/test/resources/ehr/mapper/diagnosticreport/diagnostic-report-with-coded-diagnosis.xml [hunk 1]
```diff
@@ -9,6 +9,7 @@
             <center nullFlavor="NI"/>
         </effectiveTime>
         <availabilityTime value="20100225154100"/>
+            <!-- Mapped Specimen with id: Specimen/Default-1 -->
```
Comment (mattd-kainos) [inline service/src/test/resources/ehr/mapper/diagnosticreport/diagnostic-report-with-coded-diagnosis.xml:12]: Should we still be outputting a comment / placeholder value here?

Comment (dawidskainos) [reply service/src/test/resources/ehr/mapper/diagnosticreport/diagnostic-report-with-coded-diagnosis.xml:12]: Specimen is mapped in a different class and it's mocked in DiagnosticReport tests.

File: service/src/test/resources/ehr/mapper/diagnosticreport/diagnostic-report-with-dummy-narrative.xml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        <statusCode code="COMPLETE"/>
+        <effectiveTime>
+            <center nullFlavor="NI"/>
+        </effectiveTime>
+        <availabilityTime value="20100225154100"/>
+            <!-- Mapped Specimen with id: Specimen/Default-1 -->
+            <component typeCode="COMP" contextConductionInd="true">
+    <NarrativeStatement classCode="OBS" moodCode="EVN">
+        <id root="5E496953-065B-41F2-9577-BE8F2FBD0757"/>
+        <text mediaType="text/x-h7uk-pmip">CommentType:DUMMY
```
Comment (mattd-kainos) [inline service/src/test/resources/ehr/mapper/diagnosticreport/diagnostic-report-with-dummy-narrative.xml:16]: Are we literally meant to use the text "DUMMY" in the placeholder NarrativeStatement or is there some other content that should be populated here?

Comment (mattd-kainos) [reply service/src/test/resources/ehr/mapper/diagnosticreport/diagnostic-report-with-dummy-narrative.xml:16]: There may be something in the mapping spreadsheet about this. Is row 31 relevant?

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java [hunk 3]
```diff
@@ -155,19 +157,26 @@ private String prepareReportLevelNarrativeStatements(DiagnosticReport diagnostic
...
-                diagnosticReport, comment
+                diagnosticReport, LABORATORY_RESULT_COMMENT.getCode(), comment
             );
 
             reportLevelNarrativeStatements.append(narrativeStatementFromCodedDiagnosis);
         }
 
+        if (reportLevelNarrativeStatements.length() == 0 && !diagnosticReport.hasResult()) {
+            String narrativeStatementFromCodedDiagnosis = buildNarrativeStatementForDiagnosticReport(
+                diagnosticReport, "DUMMY", EMPTY
```
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java:168]: Consider adding the additional comment type to the Enum

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java [hunk 4]
```diff
@@ -155,19 +157,26 @@ private String prepareReportLevelNarrativeStatements(DiagnosticReport diagnostic
...
             String narrativeStatementFromCodedDiagnosis = buildNarrativeStatementForDiagnosticReport(
-                diagnosticReport, comment
+                diagnosticReport, LABORATORY_RESULT_COMMENT.getCode(), comment
             );
 
             reportLevelNarrativeStatements.append(narrativeStatementFromCodedDiagnosis);
         }
 
+        if (reportLevelNarrativeStatements.length() == 0 && !diagnosticReport.hasResult()) {
+            String narrativeStatementFromCodedDiagnosis = buildNarrativeStatementForDiagnosticReport(
```
Comment (matthewdoreilly) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java:167]: From looking at the mapping spreadsheet, the dummy/default observation should look like this
```
<component typeCode="COMP" contextConductionInd="true">
    <NarrativeStatement classCode="OBS" moodCode="EVN">
        <id root="5E496953-065B-41F2-9577-BE8F2FBD0757"/>
        <text mediaType="text/x-h7uk-pmip">CommentType:AGGREGATE COMMENT SET
            CommentDate:20100225154100

            EMPTY REPORT</text>
        <statusCode code="COMPLETE"/>
        <availabilityTime value="20100225154100"/>
    </NarrativeStatement>
</component>
```

So I think `"DUMMY"` should be `CommentType.AGGREGATE_COMMENT_SET.getCode()` and `EMPTY` should be `"EMPTY REPORT"`. Also makes it clearer this is a dummy/generated narrative statement

File: service/src/main/resources/templates/diagnostic_report_compound_statement_template.mustache [hunk 5]
```diff
@@ -9,12 +9,12 @@
             <center nullFlavor="NI"/>
         </effectiveTime>
         <availabilityTime value="{{availabilityTime}}"/>
-        {{#narrativeStatements}}
-            {{{.}}}
-        {{/narrativeStatements}}
         {{#specimens}}
             {{{.}}}
         {{/specimens}}
+        {{#narrativeStatements}}
```
Comment (matthewdoreilly) [inline service/src/main/resources/templates/diagnostic_report_compound_statement_template.mustache:15]: What's the reason for moving diagnostic report level narrative statements below specimens?

Comment (dawidskainos) [reply service/src/main/resources/templates/diagnostic_report_compound_statement_template.mustache:15]: see CompundStatement schema:

```
   <xs:complexType name="RCMR_MT030101UK04.CompoundStatement">
      <xs:sequence>
         <xs:element name="id" type="II" minOccurs="1" maxOccurs="2"/>
         <xs:element name="code" type="CD" minOccurs="1" maxOccurs="1"/>
         <xs:element name="statusCode" type="CS" minOccurs="1" maxOccurs="1"/>
         <xs:element name="effectiveTime" type="IVL_TS" minOccurs="1" maxOccurs="1"/>
         <xs:element name="availabilityTime" type="TS" minOccurs="1" maxOccurs="1"/>
         <xs:element name="priorityCode" type="CV" minOccurs="0" maxOccurs="1"/>
         <xs:element name="uncertaintyCode" type="CV" minOccurs="0" maxOccurs="1"/>
         <xs:element type="RCMR_MT030101UK04.Specimen03" minOccurs="0" maxOccurs="unbounded" name="specimen"/>
         <xs:element type="RCMR_MT030101UK04.Component02" minOccurs="1" maxOccurs="unbounded" name="component"/>
         <xs:element name="informant" type="RCMR_MT030101UK04.Informant" minOccurs="0" maxOccurs="unbounded"/>
         <xs:element name="Participant" type="RCMR_MT030101UK04.Participant" minOccurs="0" maxOccurs="unbounded"/>
         <xs:element name="replacementOf" type="RCMR_MT030101UK04.ReplacementOf" minOccurs="0" maxOccurs="unbounded"/>
         <xs:element name="reason" type="RCMR_MT030101UK04.Reason" minOccurs="0" maxOccurs="unbounded"/>
         <xs:element name="reference" type="RCMR_MT030101UK04.Reference" minOccurs="0" maxOccurs="unbounded"/>
         <xs:element name="sequelTo" type="RCMR_MT030101UK04.SequelTo" minOccurs="0" maxOccurs="unbounded"/>
      </xs:sequence>
(...)
```

component element should be below specimens

File: service/src/test/resources/uat/output/TC4-9465701718_Guerra_full_20210119.xml [hunk 6]
```diff
@@ -769,7 +780,7 @@ Quantity: 1750.000 mL
         </component>
             <component typeCode="COMP" contextConductionInd="true">
     <CompoundStatement classCode="CLUSTER" moodCode="EVN">
-        <id root="6423EA0F-1F1C-4255-AEBA-D142BE836D50"/>
+        <id root="1E5D6932-C778-4997-B624-8709D2FBB189"/>
```
Comment (mattd-kainos) [inline service/src/test/resources/uat/output/TC4-9465701718_Guerra_full_20210119.xml:783]: This empty CompoundStatement fails schema validation due to missing component/specimen

org.xml.sax.SAXParseException; lineNumber: 237; columnNumber: 25; cvc-complex-type.2.4.b: The content of element 'CompoundStatement' is not complete. One of '{"urn:hl7-org:v3":specimen, "urn:hl7-org:v3":component}' is expected.

See "skipped" UAT test for Guerra. Is it related to this story or something else?

File: service/src/test/resources/ehr/mapper/diagnosticreport/diagnostic-report-with-dummy-narrative.xml (hunk 2 above)
Comment (mattd-kainos) [reply service/src/test/resources/ehr/mapper/diagnosticreport/diagnostic-report-with-dummy-narrative.xml:16]: See @matthewdoreilly comment for the correct output https://github.com/nhsconnect/integration-adaptor-gp2gp/pull/236#discussion_r637828291

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java (hunk 4 above)
Comment (matthewdoreilly) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java:167]: Also I'd put this logic in some private method called `buildDefaultNarrativeStatementForMissingObservation` or something like that

File: service/src/test/resources/uat/output/TC4-9465701718_Guerra_full_20210119.xml (hunk 6 above)
Comment (dawidskainos) [reply service/src/test/resources/uat/output/TC4-9465701718_Guerra_full_20210119.xml:783]: I think this one is covered in https://gpitbjss.atlassian.net/browse/NIAD-1470

File: service/src/test/resources/ehr/mapper/diagnosticreport/diagnostic-report-with-dummy-narrative.xml (hunk 2 above)
Comment (dawidskainos) [reply service/src/test/resources/ehr/mapper/diagnosticreport/diagnostic-report-with-dummy-narrative.xml:16]: I'll fix that according to Matthew's comment.

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java (hunk 4 above)
Comment (dawidskainos) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java:172]: done

**Review by mattd-kainos — APPROVED** [review]

**Review by Broom94 — APPROVED** [review]

**Pull request merged**